btree/btree.c

```diff
@@ -2461,8 +2465,27 @@ static bool cookie_is_valid(struct td *tree, struct m0_bcookie *k_cookie)
 	return false;
 }
 
+static int64_t lock_op_init(struct m0_sm_op *bo_op, struct node_op  *i_nop,
+			    struct td *tree, struct m0_btree_oimpl *oi, int nxt)
+{
+	/** parameters which has passed but not will be used while state machine
+	 *  implementation for  locks
+	 */
+	m0_rwlock_write_lock(&tree->t_lock);
+	oi->i_lock_aquired = true;
+	return nxt;
+}
+
+static void lock_op_unlock(struct td *tree, struct m0_btree_oimpl *oi)
+{
+	m0_rwlock_write_unlock(&tree->t_lock);
+	oi->i_lock_aquired = false;
+}
+
 static int fail(struct m0_btree_op *bop, int rc)
 {
+	if (bop->bo_i->i_lock_aquired)
+		lock_op_unlock(bop->bo_arbor->t_desc, bop->bo_i);
 	bop->bo_op.o_sm.sm_rc = rc;
 	return P_CLEANUP;
 	/* return m0_sm_op_sub(&bop->bo_op, P_CLEANUP, P_DONE); */
```

Comment:
When was this changed? Till the iteration operation implementation `return m0_sm_op_sub(&bop->bo_op, P_CLEANUP, P_DONE);`  was uncommented.

Reply:
I don't have idea, but I can change it to return m0_sm_op_sub(&bop->bo_op, P_CLEANUP, P_FINI)

Reply:
done